### ISSUE TITLE
Infinite button handler loop and watchdog bite after unmapped USB HID code.

### ISSUE THREAD

**Issue by TwinTurbine** [Hamberthm/esp32-bt2ps2#11]
Hi!

Trying to figure out how to map the unmapped keys on my Rii i4 I've found that some unmapped HID codes will send the code into an infinite loop. A single depress of an unmapped button seems to be alright but several in row seems to crash the process. A slow scroll of the scroll wheel is also alright but a "normal" scroll up or down will make the software crash. Not to mention a touch of the touchpad that will always make the software go into this infinite loop.

The output of the terminal is presented below:

```
> D (2005361) BTKeyboard: 00 00 00 00
> D (2005365) event: running post ESP_HIDH_EVENTS:2 with handler 0x4200befc and context 0x3fcad4c0 on loop 0x3fca7d78
> 0x4200befc: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621

> D (2005375) BTKeyboard: 00 00 00 00
> D (2005379) event: running post ESP_HIDH_EVENTS:2 with handler 0x4200befc and context 0x3fcad4c0 on loop 0x3fca7d78
> 0x4200befc: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621

> D (2005389) BTKeyboard: 00 00 00 00
> D (2005393) event: running post ESP_HIDH_EVENTS:2 with handler 0x4200befc and context 0x3fcad4c0 on loop 0x3fca7d78
> 0x4200befc: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621
```

And so on. Forever...

This will eventually lead to a watchdog bite.

```
D (271150) BTKeyboard: 00 00 00 00 
E (271160) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (271160) task_wdt:  - IDLE (CPU 0)
E (271160) task_wdt: Tasks currently running:
E (271160) task_wdt: CPU 0: esp_ble_hidh_ev
E (271160) task_wdt: Print CPU 0 (current core) backtrace
Print CPU 0 (current core) registers
Core  0 register dump:
MEPC    : 0x42007114  RA      : 0x42008210  SP      : 0x3fcae9d0  GP      : 0x3fc91200
0x42007114: uart_ll_get_txfifo_len at C:/Users/Project/esp/v5.2.1/esp-idf/components/hal/esp32c3/include/hal/uart_ll.h:403
 (inlined by) uart_tx_char at C:/Users/Project/esp/v5.2.1/esp-idf/components/vfs/vfs_uart.c:165
0x42008210: uart_write at C:/Users/Project/esp/v5.2.1/esp-idf/components/vfs/vfs_uart.c:210 (discriminator 2)

TP      : 0x3fc773e0  T0      : 0x00000000  T1      : 0x20000000  T2      : 0x00000000
S0/FP   : 0x00000054  S1      : 0x00000000  A0      : 0x00000000  A1      : 0x00000020
A2      : 0x00000004  A3      : 0x60000000  A4      : 0x00000001  A5      : 0xe07f8000
A6      : 0x42006972  A7      : 0x0000000a  S2      : 0x00000020  S3      : 0x00000077
0x42006972: console_write at C:/Users/Project/esp/v5.2.1/esp-idf/components/vfs/vfs_console.c:71

S4      : 0x3fc99de8  S5      : 0x3fcaeb70  S6      : 0x3c10611d  S7      : 0x00000000
S8      : 0x00000005  S9      : 0x00000006  S10     : 0x0000000b  S11     : 0x00000007
T3      : 0x00000000  T4      : 0x00000000  T5      : 0x00000000  T6      : 0x00000000
MSTATUS : 0x00000000  MTVEC   : 0x3c0e933e  MCAUSE  : 0x3fcaeaa0  MTVAL   : 0x200081d0
MHARTID : 0x00000004
D (271154) event: running post ESP_HIDH_EVENTS:2 with handler 0x4200befc and context 0x3fcad4c0 on loop 0x3fca7d78
0x4200befc: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621
```

As soon as the process recovers it will go straight back into the loop again and the only way to stop this is by removing power to the ESP as well as the keyboard. Just by removing power to the keyboard will not halt the looping, nor only removing power to the ESP to reboot it. It might work but not always.

The software seems to continue to process mapped HID codes even when in this loop, although with a noticable delay, and sometimes that can break the loop but most often not. 

If I allow myself to speculate freely, can this have something to do with a mismatch of the arrival of the HID codes from the keyboard vs the processing of the commands and the typematic output rate that leads to some kind of buffer overflow? Or is it due to the not yet fully functional mouse + wheel integration? As far as I've seen in the code there seems to be some code related to the mouse support but I'm also aware that mouse support is on the TODO list. 

For my use the touchpad/mouse support isn't critical but it would be nice to have the scroll wheel functional. I would just like to map the scroll to PgUp and PgDn and pressing of the scroll wheel to "Return". But it would be nice if a touch of the touchpad would't crash the process. :-)

Cheers!

**Comment by Hamberthm** [comment]
Hi there!

Could you please ramp up the log verbosity setting to 'VERBOSE" instead of "DEBUG" and post again a sample output chunk when you turn the wheel?

It seems the [hidh_callback()](https://github.com/Hamberthm/esp32-bt2ps2/blob/e4298b657bce084e72884f15ce24aac141fd48a8/main/bt_keyboard.cpp#L1005) is being hit constantly and that's producing the output. As far as I can think, there could be two possibilities:

1- hidh_callback() is functioning well, and the BT stack is producing the constant event callbacks. That would be outside out my reach and the BT stack is in fault and/or having conflict with your particular KB/mouse model.

2- hidh_callback() is failing at processing the event and that could be causing repeating calls (?).

Anyway, first of all we need to see what kind of events the stack is sending to the callback function, and for that we need the verbosity level turned up all the way. Then a line will be added to the output like this:

> V (1230215) BTKeyboard: f4:73:35:86:e2:5a INPUT: KEYBOARD, MAP:  0, ID:   1, Len: 8, Data:
> D (1230223) BTKeyboard: 00 00 00 00 00 00 00 00

**Comment by TwinTurbine** [comment]
Hi!

I've been trying to follow your instructions in order to change from DEBUG to VERBOSE but now I'm seeing compilation errors. I'm not sure what kind of error I introduced but this is what I get just from editing the sdconfig file.

```
[1153/1155] Linking CXX executable esp32-bt2ps2.elf
FAILED: esp32-bt2ps2.elf
cmd.exe /C "cd . && C:\Users\Project\.espressif\tools\riscv32-esp-elf\esp-13.2.0_20230928\riscv32-esp-elf\bin\riscv32-esp-elf-g++.exe -march=rv32imc_zicsr_zifencei -nostartfiles -march=rv32imc_zicsr_zifencei --specs=nosys.specs    -Wl,--cref -Wl,--defsym=IDF_TARGET_ESP32C3=0 -Wl,--Map=E:/Projekt/ESP32-Bluetooth_PS2/Github_240521/esp32-bt2ps2-v0.4-ble_ESP32C3/build/esp32-bt2ps2.map -Wl,--no-warn-rwx-segments -fno-rtti -fno-lto -Wl,--gc-sections -Wl,--warn-common -T esp32c3.peripherals.ld -T esp32c3.rom.ld -T esp32c3.rom.api.ld -T esp32c3.rom.libgcc.ld -T esp32c3.rom.newlib.ld -T esp32c3.rom.version.ld -T esp32c3.rom.eco3.ld -T memory.ld -T sections.ld @CMakeFiles\esp32-bt2ps2.elf.rsp -o esp32-bt2ps2.elf  && cd ."
C:/Users/Project/.espressif/tools/riscv32-esp-elf/esp-13.2.0_20230928/riscv32-esp-elf/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/ld.exe: esp-idf/main/libmain.a(bt_keyboard.cpp.obj): in function `BTKeyboard::ble_gap_event_handler(esp_gap_ble_cb_event_t, esp_ble_gap_cb_param_t*)':
E:/Projekt/ESP32-Bluetooth_PS2/Github_240521/esp32-bt2ps2-v0.4-ble_ESP32C3/main/bt_keyboard.cpp:791:(.text._ZN10BTKeyboard21ble_gap_event_handlerE22esp_gap_ble_cb_event_tP22esp_ble_gap_cb_param_t+0x2e8): undefined reference to `BTKeyboard::ble_gap_evt_str(unsigned char)'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

The logging section of my sdconfig looks like this:

```
# Log output
#
# CONFIG_LOG_DEFAULT_LEVEL_NONE is not set
# CONFIG_LOG_DEFAULT_LEVEL_ERROR is not set
# CONFIG_LOG_DEFAULT_LEVEL_WARN is not set
# CONFIG_LOG_DEFAULT_LEVEL_INFO is not set
# CONFIG_LOG_DEFAULT_LEVEL_DEBUG is not set
CONFIG_LOG_DEFAULT_LEVEL_VERBOSE=y
CONFIG_LOG_DEFAULT_LEVEL=5
CONFIG_LOG_MAXIMUM_EQUALS_DEFAULT=y
CONFIG_LOG_MAXIMUM_LEVEL=5
# CONFIG_LOG_MASTER_LEVEL is not set
CONFIG_LOG_COLORS=y
CONFIG_LOG_TIMESTAMP_SOURCE_RTOS=y
# CONFIG_LOG_TIMESTAMP_SOURCE_SYSTEM is not set
# end of Log output
```
It does work with DEBUG (level 4) but not with VERBOSE (level 5). I've read that VERBOSE really takes a toll on the MCU and that it might not always be available due to memory limitations. It might be due to C3 limitations so I'll try with S3 or the "classic" ESP32. Stay tuned!

**Comment by TwinTurbine** [comment]
Hi again!

I can now more ore less confirm that the VERBOSE mode isn't available on the C3. I switched back to my WROOM and it seems to work as it should. Below you'll find the output of the scroll wheel.

Scroll UP:
```
V (122311) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (122318) BTKeyboard: 00 00 00 01
D (122370) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbc64 and context 0x3ffd7808 on loop 0x3ffd3100
0x400dbc64: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621

V (122370) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (122378) BTKeyboard: 00 00 00 01
```

Scroll DOWN:
```
V (146730) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (146738) BTKeyboard: 00 00 00 ff
D (146775) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbc64 and context 0x3ffd7808 on loop 0x3ffd3100
0x400dbc64: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621

V (146775) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (146783) BTKeyboard: 00 00 00 ff
```

Scroll PUSH:
```
V (185633) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (185640) BTKeyboard: 04 00 00 00
D (185768) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbc64 and context 0x3ffd7808 on loop 0x3ffd3100
0x400dbc64: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621

V (185768) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (185775) BTKeyboard: 00 00 00 00
```

Regarding the "infinite scrolling" it is possible to generate this behaviour by what seems to be a too high input rate. If i turn the scroll wheel one "click" at a time everything is fine. But if I give it a scroll, as one might do when scrolling a web page, that will set it off.

A touch of the touch pad will immediatley set off a burst of messages like below:

```
V (54045) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (54053) BTKeyboard: 00 XX YY 00
D (54057) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbc64 and context 0x3ffd7808 on loop 0x3ffd3100
0x400dbc64: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621
```
Where XX and YY seems to be changing values according to the movement of the touch. However this will soon change to a continuous flow of tidentical messages, not responding to anything input, as far as can be seen by the eye:

```
V (54045) BTKeyboard: d1:03:45:ba:77:b8 INPUT:    MOUSE, MAP:  0, ID:   3, Len: 4, Data:
D (54053) BTKeyboard: 00 00 00 00
D (54057) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbc64 and context 0x3ffd7808 on loop 0x3ffd3100
0x400dbc64: esp_ble_hidh_event_handler_wrapper at C:/Users/Project/esp/v5.2.1/esp-idf/components/esp_hid/src/ble_hidh.c:621
```

**Comment by TwinTurbine** [comment]
Hi!

I've been doing some testing and first of all I've found some strange differences between using the Monitor Device in Visual Studio compared to using Putty when debugging. When debugging via Visual Studio the flow of messages won't even stop when the keyboard has been disconnected, even though the ESP has detected the disconnection by turning off the onboard LED. After one or more watchdog resets it might however regain control and go back to scanning for a keyboard.

Using Putty on the other hand, after going into the infinite scroll and turning off the keyboard, the scroll will stop as soon as the ESP has detected disconnection of the keyboard after 3-4 seconds (turning off the LED). It will then go straight back to scanning for the keyboard again.

I'm speculating freely now but could the scrolling has something to do with "mouse up key" input not being handled correctly, thereby not emptying any buffer? I don't know if the keyboard (mouse) expect some kind of response after the mouse up command thereby continuing to report that event? When a normal key is released ("Up key") the key passed to main is `00 00 00 00 00 00` and I haven't been able to make the software crash by pressing the normal letter keys, not even when trying to press all at once.

The software will identify pressing the Left and Right mouse button as `01 00 00 00` and `02 00 00 00` respectively and the release as `00 00 00 00`, without any fuzz. By alternately pressing the right and left button quickly the software will however start looping even though I would say that it is somewhat unlikely that the keyboard is continuing to output the release of the mouse buttons (`00 00 00 00`). But it might of course be so for some reason.

For the fun of it i tried pairing a BT mouse to the ESP generating the following report:

```
V (43489) BTKeyboard: d9:92:73:0c:e4:82 OPEN: Logi M650 L
  BLE: d9:92:73:0c:e4:82, BDA:d9:92:73:0c:e4:82, Appearance: 0x03c0, Connection ID: 0
Name: Logi M650 L, Manufacturer: Logitech, Serial Number: 3A0DF931
RSSI: -48, PID: 0xb02a, VID: 0x046d, VERSION: 0x0013
USAGE: GENERIC, Battery: Handle: 29, CCC Handle: 30
Report Maps: 1
APPEARANCE: 0x0000,   Report Map Length: 115
ADDR_TYPE: 'RANDOM',       VENDOR  OUTPUT REPORT, ID: 17, Length:  19, Permissions: 0x0e, Handle:  48, CCC Handle:   0
      VENDOR   INPUT REPORT, ID: 17, Length:  19, Permissions: 0x12, Handle:  44, CCC Handle:  45
       MOUSE   INPUT REPORT, ID:  2, Length:   7, Permissions: 0x12, Handle:  40, CCC Handle:  41
       MOUSE   INPUT   BOOT, ID:  2, Length:   3, Permissions: 0x12, Handle:  35, CCC Handle:  36
```
Moving the mouse only for brief periods the ESP seems to handle it just fine, even though the serial output will continue for some time after the movement has stopped. I guess the movement of the mouse will generate data at a faster rate than can be shifted out on the COM port. A continuous movement of the mouse will however make the watchdog bite, as with my keyboard. With the BT mouse the ESP does however seem to recover completely a while after the reset and then continue to function as one would expect.

A section of the output directly before and after the watchdog reset:
```

D (67514) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context 0x3ffd3134 on loop 0x3ffd782c
V (67524) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (67533) BTKeyboard: 00 00 09 a0 00 00 00 
D (67537) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context 0x3ffd3134 on loop 0x3ffd782c
V (67547) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (67555) BTKeyboard: 00 00 0f b0 00 00 00 
D (67560) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context E (67567) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (67567) task_wdt:  - IDLE1 (CPU 1)
E (67567) task_wdt: Tasks currently running:
E (67567) task_wdt: CPU 0: IDLE0
E (67567) task_wdt: CPU 1: esp_ble_hidh_ev
E (67567) task_wdt: Print CPU 1 backtrace


Backtrace: 0x400855EE:0x3FFBF700 0x40083ACD:0x3FFBF720 0x400D7B1A:0x3FFDB390 0x400D82FB:0x3FFDB3C0 0x400D713B:0x3FFDB3E0 0x400D6CC6:0x3FFDB400 0x4000BD83:0x3FFDB420 0x4000117D:0x3FFDB440 0x400592FE:0x3FFDB460 0x4005937A:0x3FFDB480 0x401851D1:0x3FFDB4A0 0x40196EC7:0x3FFDB4D0 0x4019688A:0x3FFDB4F0 0x4018592D:0x3FFDB810 0x401A39C9:0x3FFDB840 0x40097151:0x3FFDB870 0x401A1991:0x3FFDB8C0 0x401A1EE1:0x3FFDB900 0x401A2000:0x3FFDB950 0x40092091:0x3FFDB970

0x3ffd3134 on loop 0x3ffd782c
V (67643) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (67651) BTKeyboard: 00 00 0d 80 00 00 00 
D (67655) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context 0x3ffd3134 on loop 0x3ffd782c
V (67666) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (67674) BTKeyboard: 00 00 0f a0 00 00 00 
D (67678) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context 0x3ffd3134 on loop 0x3ffd782c
V (67688) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (67697) BTKeyboard: 00 00 12 a0 00 00 00
```
And finally, the last "mouse up" output:
```
D (77763) event: running post ESP_HIDH_EVENTS:2 with handler 0x400dbd6c and context 0x3ffd3134 on loop 0x3ffd782c
V (77763) BTKeyboard: d9:92:73:0c:e4:82 INPUT:    MOUSE, MAP:  0, ID:   2, Len: 7, Data:
D (77771) BTKeyboard: 00 00 00 00 00 00 00 
```

I don't know if this helps at all but I thought I might as well post my findings. :-)

**Comment by Hamberthm** [comment]
Hi there! Quite some work you did there! Thanks for sharing all your findings.

As far as I can tell, the mouse report is being immediately discarded at this line:

https://github.com/Hamberthm/esp32-bt2ps2/blob/e4298b657bce084e72884f15ce24aac141fd48a8/main/bt_keyboard.cpp#L1057

If the usage is not KEYBOARD (it's MOUSE as seen in your logs), then the push_key() function will not be called and the event will then be handled.

My strongest suspicion right now is that the system is so busy doing debug console outputs that some buffer (like the event queue for example) is being overflown before the callback function can process them. 

Next test you should do is turning the debug verbosity completely off, not even DEBUG level. This is to discard the system is being thrown away by outputting debug messages.

I only have one BT mouse, I'll try later to connect it later. Let me know if you can test that. If the system still crashes, then we should start looking at other things like the FreeRTOS ticking rate (I think).

**Comment by TwinTurbine** [comment]
Hi!

Well, it's was a rainy day here so I thought I might as well do something productive. :-) Regarding the debug messages it never really crossed my mind that this might be the problem in itself, as you are saying. But without these messages I'm not sure about how to detect if the problems persist. 

The code you've written seems to be rather robust so that the ESP will recover after one or more watchdog resets. As I mentioned earlier it seems that the ESP will continue to process valid key presses even while inside the infinite loop. The difference is however only that the response isn't as quick as in the "normal" case.

But of course I'll try your suggestion and see what I can find.

**Comment by Hamberthm** [comment]
Good morning,

I successfully connected my Logitech G603 gaming mouse.

- On VERBOSE log output level:
Output was fine both by turning scroll wheel and moving the mouse. Even using the 4 different sensitivity levels (reports per second), it was very hard to trigger a watchdog bite. Eventually, but very rarely, the watchdog would bite but only after grinding it hard and after many seconds.

- After turning log output to ERROR level:
On this level we should still see the watchdog messages as they are ERROR level output. I could not get one to appear, no matter how hard I drive the mouse or in what sensitivity mode. 
I even connected my Logitech K380 keyboard at the same time, and pressed some keys at the same time I was driving the mouse hard. I could see the LED blink with no delay as I pressed the key, so this confirms it was processing BT2PS2 keys as normal.

So, to me, this pretty much leads me to think you're experiencing some trouble due to the high amount of debug messages. I'm not so keen about interpreting Watchdog output, but it seems to indicate in your logs that the CPU was busy at UART code at the moment of the trigger.

As a last measure, confirm you have configTICK_RATE_HZ set to 1000 on the FreeRTOS Kernel settings in your project configuration. Just to be sure you're inline with me.

**Comment by TwinTurbine** [comment]
Hello!

I can confirm that I'm still on 1000 Hz tick rate as I haven't been fiddling with any unnecessary settings. :-) After having switched back to log level ERROR I can't seem to be able to reproduce the looping even using the touch pad or scroll wheel. For the DEBUG or VERBOSE level that was a "fail-safe" way to make the watchdog bite. So I think that your explanation for the behaviour is spot on.

**Comment by Hamberthm** [comment]
Morning! Glad to see the same results.

The loop might very possibly still be happening in the background. I strongly believe this is by design of the keyboard (touchpad), maybe to increase the touchpad's responsiveness and allow the OS to make more precise "deltas" of touch duration, by clearly identifying null touch moments. It might come out of sleep as soon as you touch it and be on until battery saving kicks in.

Up here you said the "mouse up" was also looping but in all cases it might very well be just the touchpad waking up.

Anyway, I think we can close this one and go on with the fun on #7.

Feel free to reopen if you still notice lag, resets, or watchdog bites after turning log output off.

Cheers!
